[Shawn Lee] iP

XenonShawn opened the pull request:
# Duke

| "Duke has helped me to remember important dates, 10/10 would recommend!" ~ Satisfied customer

Ever felt it was difficult to remember what you need to? Well,
Duke is here to solve all your problems! It is,

* text-based interface
* ~~trivial~~ easy to learn
* really fast!
* _free_!!

All you have to do is,

1. Download it from our [GitHub](https://github.com/XenonShawn/ip), 
2. Double click it
3. Add your tasks
4. And let it manage your tasks for you! 😉

Did we also mention that its **FREE**?

Features:
- [x] Managing tasks
- [ ] Managing deadlines
- [ ] Reminders
-----
If you are a programmer, you can also use it to practice Java. Here is the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

XenonShawn committed:
Skeletal version of Duke - simply echos commands until 'bye' is received

XenonShawn committed:
Level-2

XenonShawn committed:
Level-3

XenonShawn committed:
Level-4

XenonShawn committed:
Level-5

XenonShawn committed:
Level-6

XenonShawn committed:
Exceptions

XenonShawn committed:
Add save feature

Tasks created are not saved to disk.

Allowing saving to disk aids greatly in usability.

XenonShawn committed:
Add date parsing functionality for `Deadline` and `Event` tasks

The due date for both was stored as a String.

Storing it as an actual date will allow for more automated features to
work on it in the future.

XenonShawn committed:
Merge branch 'branch-Level-7'

XenonShawn committed:
Merge branch 'branch-Level-8'

XenonShawn committed:
Merge remote-tracking branch 'upstream/add-gradle-support' into add-gradle-support

XenonShawn committed:
Merge branch 'add-gradle-support'

XenonShawn committed:
Refactor the code to include more OOP

XenonShawn committed:
Add JUnit tests for some files

XenonShawn committed:
Add JavaDoc comments

XenonShawn committed:
Add Find functionality

XenonShawn committed:
Merge branch 'branch-Level-9'

XenonShawn committed:
Add JavaDoc message for FindCommand

shaniceng commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,152 @@
+package duke.parser;
+
+import duke.commands.*;
```
Suggest to not import all in this manner and state each import statement explicitly

shaniceng commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,152 @@
+        int idx = input.indexOf(' ');
+        String textCommand = idx < 0 ? input : input.substring(0, idx);
+
+        switch (textCommand) {
```
Great job in having the switch statement written in the correct indentations as well as keeping it neat with all the necessary work done in the private methods below👍

crvstalphua replied:
Agreed, I think it was great that your case and switch indentations are correct!

shaniceng commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,96 @@
+     * @return Tasks stored on disk.
+     */
+    public ArrayList<Task> load() {
+        ArrayList<Task> tasks = new ArrayList<>();
```
Great job in naming an array list as plural 💯

shaniceng commented on src/main/java/duke/tasks/Task.java:
```diff
@@ -0,0 +1,70 @@
+    }
+
+    /**
+     * Mark the task as completed.
```
Suggest to change mark to marks instead 
```suggestion
     * Marks the task as completed.
```

shaniceng commented on src/main/java/duke/tasks/Task.java:
```diff
@@ -0,0 +1,70 @@
+    }
+
+    /**
+     * Unmark the task, ie not completed.
```
Suggest to change from unmark to unmarks
```suggestion
     * Unmarks the task, ie not completed.
```

shaniceng commented on src/main/java/duke/tasks/TaskList.java:
```diff
@@ -0,0 +1,36 @@
+        this.storage = storage;
+        this.tasks = storage.load();
+    }
+
```
Great job in naming all of the methods in verbs and in camel cases 👍

shaniceng commented on src/main/java/duke/commands/ByeCommand.java:
```diff
@@ -0,0 +1,30 @@
+    /**
+     * Indication to stop the bot.
+     *
+     * @return true - A signal to stop the bot.
```
Great job in giving the methods comments, but suggest to state what does true means. For eg: 
```suggestion
     * @return true If user stops the bot.
```

shaniceng reviewed (commented):
Overall, your code is easy to read and it LGTM, just some nits to be fixed 👍

chngchngchng commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,40 @@
+import duke.storage.Storage;
+import duke.tasks.TaskList;
+import duke.ui.Ui;
+
```
Perhaps add javadocs to public classes?

chngchngchng commented on src/main/java/duke/commands/ByeCommand.java:
```diff
@@ -0,0 +1,30 @@
+import duke.storage.Storage;
+import duke.tasks.TaskList;
+import duke.ui.Ui;
+
```
Perhaps add javadocs to public classes?

crvstalphua commented on src/main/java/duke/tasks/Event.java:
```diff
@@ -0,0 +1,37 @@
+package duke.tasks;
+
+import java.time.LocalDate;
+import java.time.format.DateTimeFormatter;
+import java.time.format.DateTimeParseException;
+
+import duke.exceptions.DukeException;
+
```
Suggest to have a constant order for import statements. Perhaps, the DukeException statement could be above the import time statements (as in Deadline class)?

```suggestion
import duke.exceptions.DukeException;

import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.time.format.DateTimeParseException;

```

chngchngchng commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,96 @@
+                    default:
+                        continue;
+                    }
+                } catch (DukeException e) {
+                    // Invalid data, just ignore the row
+                    continue;
+                }
```
Perhaps change behaviour of Duke to overwrite all data in the file if invalid data is provided?

chngchngchng commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,96 @@
+     * @param tasks The <Code>ArrayList</Code> of tasks.
+     */
+    public void update(ArrayList<Task> tasks) {
+        StringBuilder store = new StringBuilder();
+
+        for (Task task : tasks) {
+            store.append(task.encode()).append("\n");
+        }
+
+        try {
+            FileWriter writer = new FileWriter(path);
+            writer.write(store.toString());
+            writer.close();
+        } catch (IOException e) {
+            System.out.println("Unable to store your files :(");
+        }
```
Good use of String Builder!

chngchngchng reviewed (commented):
In general, code structure is solid and LGTM! Just remember to add javadoc comments to public classes and methods.

crvstalphua commented on src/main/java/duke/commands/Command.java:
```diff
@@ -0,0 +1,15 @@
+
+    public abstract void execute(TaskList tasks, Ui ui, Storage storage) throws DukeException;
+
+    public boolean isExit() {
```
Great that your boolean variable is named to sound like a boolean!

crvstalphua commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,45 @@
+        System.out.println(DIVIDER);
+    }
+
+    public Command readCommand() throws DukeException {
```
Great that all your method names are in camelCase and are verbs as well!

crvstalphua commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,40 @@
+package duke;
```
Great that all your classes are in packages!

crvstalphua reviewed (commented):
Overall, the code was formatted really neatly, making it easier to read! Many of the methods had header comments following the javadoc format, but would be good if you could add header comments for all the other methods as well. LGTM, just some minor potential edits :D

XenonShawn committed:
Add Checkstyle to detect coding style violations

XenonShawn committed:
Add a graphical user interface

Textual interfaces are not clean.

Use JavaFx to create a new GUI to make the application more user
friendly.

Ui class has been refactored into a TextUi class which inherits
from a new Ui interface. Duke has been refactor to take into account
this change.

XenonShawn committed:
Fix a minor style issue in Main.java

XenonShawn committed:
Merge branch 'branch-Level-10'

XenonShawn committed:
Update gradle build file for new JAR file

XenonShawn committed:
Update commands to use StringBuilder

XenonShawn committed:
Introduce assertions to Duke

Duke.java makes assumptions on state.

Introducing assertions to make those assumptions clear will improve code
clarity.

XenonShawn committed:
Update JavaDocs to better fit coding standard

XenonShawn committed:
Update Javadocs to better fit coding standard

XenonShawn committed:
Add and update Javadocs to improve code quality

XenonShawn committed:
Merge branch 'branch-A-CodeQuality'

XenonShawn committed:
Add some shorter aliases for convenience

XenonShawn committed:
Add in command aliasing

Commands have to be exact.

Adding command aliasing is more user-friendly as it allows users to
use whatever commands they prefer.

XenonShawn committed:
Solve bug regarding deletion of tasks indexed 1

XenonShawn committed:
Add product screenshot

XenonShawn committed:
Merge branch 'branch-A-CodeQuality'

XenonShawn committed:
Fix bug regarding e being an alias for deadline instead of event

XenonShawn committed:
Add a brief User Guide

XenonShawn committed:
Merge pull request #2 from XenonShawn/master

Satisfy Week 5 Merging PR requirement

XenonShawn committed:
Merge pull request #3 from XenonShawn/branch-A-CodeQuality

Merge pull request #2 from XenonShawn/master

XenonShawn committed:
Fix bye command for GUIs
